pegasus/helpers/hoc_helpers.rb

```diff
@@ -41,9 +88,10 @@ def set_hour_of_code_cookie_for_row(row)
 
 def complete_tutorial(tutorial={})
   unless settings.read_only
-    # We intentionally allow this DB write even when hoc_activity_writes_disabled
-    # is set so we can generate personalized, shareable certificates.
-    row = DB[:hoc_activity].where(session: request.cookies['hour_of_code']).first
+    # We intentionally allow this DB write even for otherwise unsampled sessions so we can
+    # generate personalized, shareable certificates. Only a fraction of users reach the end
+    # so its OK to write here.
+    row = DB[:hoc_activity].where(session: session_id).first
```

Comment:
Heh, github decided to collapse lines 98-103 below.  I knew there was an else branch!

If you expand it you'll see that the required else clause is there.